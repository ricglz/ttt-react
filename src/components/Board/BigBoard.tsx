import React from 'react';
import type { BigBoard as BigBoardType, CurrentBoard } from '../../functions/HelperFunctions';
import BoardRow from './BoardRow';

type Props = {
  boardGame: BigBoardType,
  currentBoard: CurrentBoard,
  handleClick: (board: number, id: number) => void
};

const BigBoard = ({ boardGame, currentBoard, handleClick }: Props) => (
  <div className="game row justify-content-center">
    <div className="col-9 col-sm-8 col-md-7 col-lg-6 col-xl-6">
      <BoardRow
        rowNum={0}
        boardGame={boardGame}
        currentBoard={currentBoard}
        handleClick={handleClick}
      />
      <BoardRow
        rowNum={3}
        boardGame={boardGame}
        currentBoard={currentBoard}
        handleClick={handleClick}
      />
      <BoardRow
        rowNum={6}
        boardGame={boardGame}
        currentBoard={currentBoard}
        handleClick={handleClick}
      />
    </div>
  </div>
);

export default BigBoard;
