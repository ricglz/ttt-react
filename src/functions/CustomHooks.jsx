import React from 'react';
import {
  isOccupied,
  theresAWinner,
  initialState,
  alertWinner,
} from './HelperFunctions';
import makeMove, {
  cleanVariables,
  aiMadeAMove,
} from './Ai';

const CONSTANTS = {
  PLAYER1: 'X',
  PLAYER2: 'O',
};

export function useGameHooks(game, setGame) {
  const {
    boardGame, currentBoard, currentPlayer,
  } = game;

  const canClick = React.useCallback((board, id) => {
    const currentValue = boardGame[board][id];
    if (isOccupied(currentValue)) return false;
    return board === currentBoard || currentBoard === -1;
  }, [boardGame, currentBoard]);

  const pvpMove = React.useCallback((boardCopy, newMoveNumber, id) => {
    setGame({
      ...game,
      boardGame: boardCopy,
      moveNumber: newMoveNumber,
      currentBoard: id,
      currentPlayer:
        currentPlayer === CONSTANTS.PLAYER1
          ? CONSTANTS.PLAYER2
          : CONSTANTS.PLAYER1,
    });
  }, [setGame, game, currentPlayer]);

  const newGame = React.useCallback(() => {
    cleanVariables();
    setGame({ ...game, ...initialState() });
  }, [setGame, game]);

  return [canClick, pvpMove, newGame];
}

export function useAIHooks({
  moveNumber,
  selectedOption,
  setGame,
  changeScore,
  newGame,
  setSelectedOption,
  game,
}) {
  const aiMove = React.useCallback((boardCopy, id, newMoveNumber) => {
    const board = boardCopy;
    const difficulty = selectedOption == null ? 1 : selectedOption.value;
    const move = makeMove(board[id], id, difficulty);
    board[id][move] = -1;
    const winner = theresAWinner(board[id]);
    if (winner) {
      alertWinner(winner);
      changeScore(winner);
      newGame();
    } else {
      aiMadeAMove(id, board[id]);
      setGame({
        ...game,
        boardGame: board,
        moveNumber: newMoveNumber + 1,
        currentBoard: move,
      });
    }
  }, [changeScore, newGame, selectedOption, setGame, game]);

  const handleChange = React.useCallback((newSelectedOption) => {
    setSelectedOption(newSelectedOption);
    if (moveNumber >= 0) {
      newGame();
    }
  }, [setSelectedOption, moveNumber, newGame]);

  return [aiMove, handleChange];
}
