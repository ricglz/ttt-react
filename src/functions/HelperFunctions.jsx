import { NotificationManager } from 'react-notifications';

let boardCopy = null;

export function alertWinner(winner) {
  const winnerLabel = winner === 1 ? 'X' : 'O';
  NotificationManager.info(`Player ${winnerLabel} has won.`);
}

export function emptyArray() {
  return [0, 0, 0, 0, 0, 0, 0, 0, 0];
}

export function isOccupied(value) {
  return value !== 0;
}

function allThree(first, second, third) {
  const firstValue = boardCopy[first];
  const secondValue = boardCopy[second];
  const thirdValue = boardCopy[third];
  if (
    firstValue === secondValue
    && secondValue === thirdValue
    && isOccupied(firstValue)
  ) {
    return firstValue;
  }
  return null;
}

function columnWin() {
  return allThree(0, 3, 6) || allThree(1, 4, 7) || allThree(2, 5, 8);
}

function rowWin() {
  return allThree(0, 1, 2) || allThree(3, 4, 5) || allThree(6, 7, 8);
}

function diagonalWin() {
  return allThree(0, 4, 8) || allThree(2, 4, 6);
}

export function theresAWinner(board) {
  boardCopy = board;
  return columnWin() || rowWin() || diagonalWin();
}

export function newBoard() {
  return [
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
  ];
}

export function initialState() {
  return {
    boardGame: newBoard(),
    currentPlayer: 'X',
    moveNumber: 0,
    currentBoard: -1,
  };
}

export function fbInitialState(hostUid, hostName) {
  return {
    boardGame: newBoard(),
    currentPlayer: 'X',
    moveNumber: 0,
    currentBoard: -1,
    oWins: 0,
    xWins: 0,
    hostUid,
    hostName,
    guestUid: -1,
    nextPlayerUid: hostUid,
    timestamp: Date.now(),
  };
}
