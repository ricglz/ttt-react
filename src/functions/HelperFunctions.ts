import { NotificationManager } from 'react-notifications';
import type {
  BigBoard, Board, GeneralBoardIndex, NumberBoard,
} from '../@types/general';
import { Cell, Player } from '../@types/general_enums';

export const getNextPlayer = (player: Player) => (
  player === Player.PLAYER_1 ? Player.PLAYER_2 : Player.PLAYER_1
);

export const getPlayerCellValue = (player: Player) => (
  player === Player.PLAYER_1 ? Cell.X : Cell.O
);

export function alertWinner(winner: Cell) {
  const winnerLabel = winner === Cell.X ? Player.PLAYER_1 : Player.PLAYER_2;
  NotificationManager.info(`Player ${winnerLabel} has won.`);
}

export function alertError(err: Error) {
  NotificationManager.error(err.message);
}

export function emptyArray(): NumberBoard {
  return [0, 0, 0, 0, 0, 0, 0, 0, 0];
}

export function isOccupied(value: Cell) {
  return value !== Cell.NONE;
}

function allThree(
  first: GeneralBoardIndex,
  second: GeneralBoardIndex,
  third: GeneralBoardIndex,
  board: Board,
) {
  const firstValue = board[first];
  const secondValue = board[second];
  const thirdValue = board[third];
  if (
    firstValue === secondValue
    && secondValue === thirdValue
    && isOccupied(firstValue)
  ) {
    return firstValue;
  }
  return null;
}

function columnWin(board: Board) {
  return allThree(0, 3, 6, board) || allThree(1, 4, 7, board) || allThree(2, 5, 8, board);
}

function rowWin(board: Board) {
  return allThree(0, 1, 2, board) || allThree(3, 4, 5, board) || allThree(6, 7, 8, board);
}

function diagonalWin(board: Board) {
  return allThree(0, 4, 8, board) || allThree(2, 4, 6, board);
}

export function theresAWinner(board: Board) {
  return columnWin(board) || rowWin(board) || diagonalWin(board);
}

export function newBoard(): BigBoard {
  return [
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 0, 0],
  ];
}

export type CurrentBoard = GeneralBoardIndex | -1;

interface BaseGame {
  boardGame: BigBoard;
  currentPlayer: Player;
  moveNumber: number;
  currentBoard: CurrentBoard;
}

export function initialState(): BaseGame {
  return {
    boardGame: newBoard(),
    currentPlayer: Player.PLAYER_1,
    moveNumber: 0,
    currentBoard: -1,
  };
}

export interface FirebaseGame extends BaseGame {
  oWins: number;
  xWins: number;
  hostUid: string;
  hostName: string;
  guestUid: string;
  nextPlayerUid: string;
  timestamp: number;
}

export function fbInitialState(hostUid: string, hostName: string): FirebaseGame {
  return {
    ...initialState(),
    oWins: 0,
    xWins: 0,
    hostUid,
    hostName,
    guestUid: '-1',
    nextPlayerUid: hostUid,
    timestamp: Date.now(),
  };
}

export type Game = BaseGame | FirebaseGame;
