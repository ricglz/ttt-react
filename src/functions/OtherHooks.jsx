import React from 'react';
import { getRedirect } from '../firebase/firebase';

function updateUser(setUser) {
  getRedirect().then((response) => {
    const tempUser = response.user;
    if (!tempUser) {
      return;
    }
    const {
      displayName, email, phoneNumber, photoUrl, uid,
    } = tempUser;
    const user = {
      name: displayName, email, phoneNumber, photoUrl, uid,
    };
    setUser(user);
    localStorage.setItem('user', JSON.stringify(user));
  });
}

// eslint-disable-next-line import/prefer-default-export
export function useUser() {
  const cachedUser = localStorage.getItem('user');
  const initialUser = cachedUser ? JSON.parse(cachedUser) : {};
  const [user, setUser] = React.useState(initialUser);

  React.useEffect(() => {
    updateUser(setUser);
  }, [setUser]);

  const logOut = React.useCallback(() => {
    setUser({});
    localStorage.removeItem('user');
  }, [setUser]);

  return [user, logOut];
}
