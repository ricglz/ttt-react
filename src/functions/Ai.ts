import type { Board, GeneralBoardIndex, NumberBoard } from "types/general";

import AiAction from "./AiAction";
import { Cell, Difficulty } from "types/general_enums";
import { isOccupied, emptyArray } from "./HelperFunctions";

interface Args {
  board: Board;
  currentBoard: GeneralBoardIndex;
  currentDifficulty: Difficulty;
}

interface AiProperties extends Args {
  amountOccupied: NumberBoard;
  avoidBox: NumberBoard;
}

function areTwoValue(value: Cell) {
  if (value === Cell.X) return 90;
  return 40;
}

// Checks which are the more factible actions to play
function different(availableMoves: AiAction[]) {
  const min = availableMoves[0].value;
  return availableMoves.findIndex((action) => action.value > min);
}

function deleteElements(availableMoves: AiAction[]) {
  availableMoves.sort((a, b) => b.value - a.value);
  const erasePos = different(availableMoves);
  if (erasePos !== -1) {
    availableMoves.splice(erasePos, availableMoves.length - erasePos);
  }
  return availableMoves;
}

function getAvailableMoves(props: AiProperties) {
  const availableMoves: AiAction[] = [];
  props.board.forEach((element, index) => {
    if (isOccupied(element)) {
      return;
    }
    const pos = index as GeneralBoardIndex;
    availableMoves.push(
      new AiAction({ pos, boardCopy: props.board, ...props })
    );
  });
  return availableMoves;
}

class Ai {
  board: Board;

  currentBoard: GeneralBoardIndex;

  currentDifficulty: Difficulty;

  amountOccupied = emptyArray();

  avoidBox = emptyArray();

  constructor({ board, currentBoard, currentDifficulty }: Args) {
    this.board = board;
    this.currentDifficulty = currentDifficulty;
    this.currentBoard = currentBoard;
  }

  setVariables = ({ board, currentBoard, currentDifficulty }: Args) => {
    this.board = board;
    this.currentDifficulty = currentDifficulty;
    this.currentBoard = currentBoard;
  };

  getProperties = (): AiProperties => ({
    amountOccupied: this.amountOccupied,
    avoidBox: this.avoidBox,
    board: this.board,
    currentBoard: this.currentBoard,
    currentDifficulty: this.currentDifficulty,
  });

  areTwo = (
    pos1: GeneralBoardIndex,
    pos2: GeneralBoardIndex,
    pos3: GeneralBoardIndex
  ) => {
    const value1 = this.board[pos1];
    const value2 = this.board[pos2];
    const value3 = this.board[pos3];

    const firstIsOccupied = isOccupied(value1);
    const secondIsOccupied = isOccupied(value2);

    if (
      (value1 === value2 && firstIsOccupied && !isOccupied(value3)) ||
      (value1 === value3 && firstIsOccupied && !secondIsOccupied)
    ) {
      return areTwoValue(value1);
    }
    if (value2 === value3 && secondIsOccupied && !firstIsOccupied) {
      return areTwoValue(value2);
    }
    return 0;
  };

  areTwoInTheColumn = () =>
    this.areTwo(0, 3, 6) + this.areTwo(1, 4, 7) + this.areTwo(2, 5, 8);

  areTwoInTheRow = () =>
    this.areTwo(0, 1, 2) + this.areTwo(3, 4, 5) + this.areTwo(6, 7, 8);

  areTwoInTheDiagonal = () => this.areTwo(0, 4, 8) + this.areTwo(2, 4, 6);

  areTwoInTheBoard = () =>
    this.areTwoInTheColumn() +
    this.areTwoInTheDiagonal() +
    this.areTwoInTheRow();

  playerMadeAMove = (boardId: GeneralBoardIndex, board: Board) => {
    this.amountOccupied[boardId] += 3;
    this.board = board;
    this.avoidBox[boardId] = this.areTwoInTheBoard();
  };

  aiMadeAMove = (board: Board) => {
    this.amountOccupied[this.currentBoard] += 1;
    this.board = board;
    this.avoidBox[this.currentBoard] = this.areTwoInTheBoard();
  };

  cleanVariables = () => {
    this.amountOccupied = emptyArray();
    this.avoidBox = emptyArray();
  };

  makeMove = (args: Args) => {
    this.setVariables(args);
    const properties = this.getProperties();
    let availableMoves = getAvailableMoves(properties);
    if (this.currentDifficulty >= Difficulty.NORMAL) {
      availableMoves = deleteElements(availableMoves);
    }
    const index = Math.floor(Math.random() * availableMoves.length);
    return availableMoves[index].pos;
  };
}

export default Ai;
