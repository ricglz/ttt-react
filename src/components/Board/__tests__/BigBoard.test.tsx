import renderer from 'react-test-renderer';
import { newBoard } from '../../../functions/HelperFunctions';
import BigBoard from '../BigBoard';

it('renders without crashing', () => {
  const boardGame = newBoard();
  const component = (
    <BigBoard
      boardGame={boardGame}
      currentBoard={-1}
      handleClick={() => {}}
    />
  );
  const tree = renderer.create(component).toJSON();
  expect(tree).toMatchSnapshot();
});
