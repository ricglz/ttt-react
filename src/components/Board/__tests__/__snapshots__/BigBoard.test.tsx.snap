// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<div
  className="game row justify-content-center"
>
  <div
    className="col-9 col-sm-8 col-md-7 col-lg-6 col-xl-6"
  >
    <div
      className="game row"
    >
      <div
        className="col-4 big-box"
      >
        <div
          className="game row cell-row"
        >
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
        </div>
        <div
          className="game row cell-row"
        >
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
        </div>
        <div
          className="game row cell-row"
        >
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
        </div>
      </div>
      <div
        className="col-4 big-box"
      >
        <div
          className="game row cell-row"
        >
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
        </div>
        <div
          className="game row cell-row"
        >
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
        </div>
        <div
          className="game row cell-row"
        >
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
        </div>
      </div>
      <div
        className="col-4 big-box"
      >
        <div
          className="game row cell-row"
        >
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
        </div>
        <div
          className="game row cell-row"
        >
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
        </div>
        <div
          className="game row cell-row"
        >
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
        </div>
      </div>
    </div>
    <div
      className="game row"
    >
      <div
        className="col-4 big-box"
      >
        <div
          className="game row cell-row"
        >
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
        </div>
        <div
          className="game row cell-row"
        >
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
        </div>
        <div
          className="game row cell-row"
        >
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
        </div>
      </div>
      <div
        className="col-4 big-box"
      >
        <div
          className="game row cell-row"
        >
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
        </div>
        <div
          className="game row cell-row"
        >
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
        </div>
        <div
          className="game row cell-row"
        >
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
        </div>
      </div>
      <div
        className="col-4 big-box"
      >
        <div
          className="game row cell-row"
        >
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
        </div>
        <div
          className="game row cell-row"
        >
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
        </div>
        <div
          className="game row cell-row"
        >
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
        </div>
      </div>
    </div>
    <div
      className="game row"
    >
      <div
        className="col-4 big-box"
      >
        <div
          className="game row cell-row"
        >
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
        </div>
        <div
          className="game row cell-row"
        >
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
        </div>
        <div
          className="game row cell-row"
        >
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
        </div>
      </div>
      <div
        className="col-4 big-box"
      >
        <div
          className="game row cell-row"
        >
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
        </div>
        <div
          className="game row cell-row"
        >
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
        </div>
        <div
          className="game row cell-row"
        >
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
        </div>
      </div>
      <div
        className="col-4 big-box"
      >
        <div
          className="game row cell-row"
        >
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
        </div>
        <div
          className="game row cell-row"
        >
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
        </div>
        <div
          className="game row cell-row"
        >
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
          <div
            className="col-4 box"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={-1}
          >
            <p
              className=""
            >
              
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
